feat(client-core-interfaces): `JsonString`

jason-ha opened the pull request:
`JsonString<T>` represents a string that has a type `T` encoded within.
If a value of type `T` were given to `JSON.stringify` (and is serializable), the string produced is `JsonString<T>`.
That same string passed to `JSON.parse`, will produce a value of type `JsonDeserialized<T>`. (Value may not be of type `T` because there may be loss during stringify. `JsonDeserialized` filter is very accurate representation of the result.)

`JsonStringify` and `JsonParse` utilities are provided that output and take-in `JsonString` strings using regular `JSON.stringify` and `JSON.parse`, respectively. `JsonStringify` employs `JsonSerializable` type filter on input to reveal most serialization compatibility issues.

Test cases added:
```text
  JsonString
    ✔ `JsonString<string>` is assignable to `string`
    ✔ `string` is not assignable to `JsonString<unknown>`
    ✔ object is not assignable to `JsonString<unknown>`
    ✔ `JsonString<A | B>` is assignable to `JsonString<A> | JsonString<B>` without enums involved
    ✔ `JsonString<A | B>` is assignable to `JsonString<A> | JsonString<B>` with enums involved
    is covariant over T
      ✔ `JsonString<"literal">` is assignable to `JsonString<string>`
      ✔ `JsonString<Record<string, number>>` is assignable to `JsonString<Record<string, number | undefined>>`
      ✔ `JsonString<Record<string, number | undefined>>` is assignable to `JsonString<Record<string, unknown>>`
      ✔ `JsonString<bigint>` is assignable to `JsonString<unknown>`
      ✔ `JsonString<A> | JsonString<B>` is assignable to `JsonString<A | B>`
    is not contravariant over T
      ✔ `JsonString<string>` is not assignable to `JsonString<"literal">`
      ✔ `JsonString<Record<string, number | undefined>>` is not assignable to `JsonString<Record<string, number>>`
      ✔ `JsonString<unknown>` is not assignable to `JsonString<bigint>`
      ✔ `JsonString<A | B>` is not assignable to `JsonString<A>`

  JsonStringify and JsonParse
    expected usage
      supports primitive types
        ✔ `boolean`
        ...
        ✔ branded `string`
        ✔ `JsonString<string>`
        ✔ `JsonString<{ arrayOfNumbers: number[] }>`
        ✔ `JsonString<Record<string, number>>`
        ✔ `JsonString<Record<string, number | undefined>>`
        ✔ `JsonString<bigint>`
        ✔ `JsonString<unknown>`
      supports literal types
        ✔ `true`
        ✔ `false`
        ✔ `0`
        ✔ "string"
        ✔ `null`
        ✔ object with literals
        ✔ array of literals
        ✔ tuple of literals
        ✔ specific numeric enum value
        ✔ specific string enum value
        ✔ specific const heterogenous enum value
        ✔ specific computed enum value
      supports array types
        ✔ array of `number`s
        ✔ readonly array of `number`s
        ✔ readonly array of simple objects
      supports object types
        ✔ empty object
        ✔ object with `never`
        ✔ object with `boolean`
        ...
        ✔ `string`|`number` indexed record of `string`s
        ✔ `string`|`number` indexed record of objects
        ✔ templated record of `numbers`
        ✔ `string` indexed record of `number`|`string`s with known properties
        ✔ `string`|`number` indexed record of `strings` with known `number` property (unassignable)
        ✔ branded-`string` indexed of `boolean`s
        ...
        ✔ object with possible type recursion through union
        ✔ object with optional type recursion
        ✔ object with deep type recursion
        ✔ object with alternating type recursion
        ✔ simple non-null object json (NonNullJsonObjectWith<never>)
        ✔ simple read-only non-null object json (ReadonlyNonNullJsonObjectWith<never>)
        ✔ non-const enums
        ✔ object with `readonly`
        ✔ object with getter implemented via value
        ✔ object with setter implemented via value
        ✔ object with matched getter and setter implemented via value
        ✔ object with mismatched getter and setter implemented via value
        class instance (losing 'instanceof' nature)
          ✔ with public data (just cares about data)
          with `ignore-inaccessible-members`
            ✔ with private method ignores method
            ✔ with private getter ignores getter
            ✔ with private setter ignores setter
        object with optional property
          ✔ without property
          ✔ with undefined value
          ✔ with defined value
        opaque Json types
          ✔ opaque serializable object
          ✔ opaque deserialized object
          ✔ opaque serializable and deserialized object
          ✔ opaque serializable unknown
          ✔ opaque deserialized unknown
          ✔ opaque serializable and deserialized unknown
          ✔ object with opaque serializable unknown
          ✔ object with opaque deserialized unknown
          ✔ recursive type with opaque serializable unknown
          ✔ recursive type with opaque deserialized unknown
          ✔ recursive type with opaque serializable and deserialized unknown
          ✔ recursive branded indexed object with OpaqueJsonDeserialized<unknown>
      supports union types
        ✔ simple json (JsonTypeWith<never>)
        ✔ simple read-only json (ReadonlyJsonTypeWith<never>)
      with NOT fully supported object types
        ✔ object with self reference throws on serialization
        known defect expectations
          ✔ sparse array of supported types
          ✔ object with sparse array of supported types
          getters and setters allowed but do not propagate
            ✔ object with `readonly` implemented via getter
            ✔ object with getter
            ✔ object with setter
            ✔ object with matched getter and setter
            ✔ object with mismatched getter and setter
          class instance
            with `ignore-inaccessible-members`
              ✔ with private data ignores private data (that propagates)
    invalid input usage
      assumptions
        ✔ const enums are never readable
      unsupported types cause compiler error
        ✔ `undefined`
        ✔ `unknown`
        ✔ `symbol`
        ✔ `unique symbol`
        ✔ `bigint`
        ✔ function
        ✔ function with supported properties
        ✔ object and function
        ✔ object with function with supported properties
        ✔ object with object and function
        ✔ function with class instance with private data
        ✔ function with class instance with public data
        ✔ class instance with private data and is function
        ✔ class instance with public data and is function
        ✔ `object` (plain object)
        ✔ `void`
        ✔ branded `object`
        ✔ branded object with `string`
        unions with unsupported primitive types
          ✔ `string | symbol`
          ✔ `bigint | string`
          ✔ `bigint | symbol`
          ✔ `number | bigint | symbol`
        array
          ✔ array of `bigint`s
          ✔ array of `symbol`s
          ...
        object
          ✔ object with exactly `bigint`
          ✔ object with optional `bigint`
          ✔ object with exactly `symbol`
          ...
          ✔ `string` indexed record of `unknown`
          ✔ `Partial<>` `string` indexed record of `unknown`
          ✔ `Partial<>` `string` indexed record of `numbers`
          ✔ `Partial<>` templated record of `numbers`
          ✔ object with recursion and `symbol`
          ✔ function object with recursion
          ✔ object and function with recursion
          ✔ nested function object with recursion
          ✔ nested object and function with recursion
          ✔ object with inherited recursion extended with unsupported properties
          object with `undefined`
            ✔ as exact property type
            ✔ in union property
            ...
          object with required `unknown`
            ✔ as exact property type
            ✔ as exact property type adjacent to recursion
            ✔ as exact property type in recursion
          of class instance
            ✔ with private data
            ✔ with private method
            ✔ with private getter
            ✔ with private setter
            ✔ with public method
            ✔ with private data in optional recursion
        opaque Json types requiring extra allowed types
          ✔ opaque serializable object with `bigint`
          ✔ opaque deserialized object with `bigint`
          ✔ opaque serializable and deserialized object with `bigint`
          ✔ opaque serializable object with number array expecting `bigint` support
          ✔ opaque deserialized object with number array expecting `bigint` support
          ✔ opaque serializable and deserialized object with number array expecting `bigint` support
        common class instances
          ✔ Map
          ✔ ReadonlyMap
          ✔ Set
          ✔ ReadonlySet
        Fluid types
          ✔ `IFluidHandle`
          ✔ object with `IFluidHandle`
    special cases
      ✔ explicit `any` generic still limits allowed types
      `number` edge cases
        supported
          ✔ MIN_SAFE_INTEGER
          ✔ MAX_SAFE_INTEGER
          ✔ MIN_VALUE
          ✔ MAX_VALUE
        resulting in `null`
          ✔ NaN
          ✔ +Infinity
          ✔ -Infinity

  JsonParse
    ✔ parses `JsonString<A> | JsonString<B>` to `JsonDeserialized<A | B>`

  JsonDeserialized
    positive compilation tests
      supported primitive types are preserved
        ✔ `JsonString<string>`
        ✔ `JsonString<{ arrayOfNumbers: number[] }>`
        ✔ `JsonString<Record<string, number>>`
        ✔ `JsonString<Record<string, number | undefined>>`
        ✔ JsonString<bigint>
        ✔ JsonString<unknown>

  JsonSerializable
    positive compilation tests
      supported primitive types
        ✔ `JsonString<string>`
        ✔ `JsonString<{ arrayOfNumbers: number[] }>`
        ✔ `JsonString<Record<string, number>>`
        ✔ `JsonString<Record<string, number | undefined>>`
        ✔ JsonString<bigint>
        ✔ JsonString<unknown>
```

jason-ha committed:
feat(client-core-interfaces): `JsonString`
with `JsonStringify` and `JsonParse` utilities

Test cases added:
```text
  JsonString
    ✔ `JsonString<string>` is assignable to `string`
    ✔ `string` is not assignable to `JsonString<unknown>`
    ✔ object is not assignable to `JsonString<unknown>`
    is covariant over T
      ✔ `JsonString<"literal">` is assignable to `JsonString<string>`
      ✔ `JsonString<Record<string, number>>` is assignable to `JsonString<Record<string, number | undefined>>`
      ✔ `JsonString<Record<string, number | undefined>>` is assignable to `JsonString<Record<string, unknown>>`
      ✔ `JsonString<bigint>` is assignable to `JsonString<unknown>`
    is not contravariant over T
      ✔ `JsonString<string>` is not assignable to `JsonString<"literal">`
      ✔ `JsonString<Record<string, number | undefined>>` is not assignable to `JsonString<Record<string, number>>`
      ✔ `JsonString<unknown>` is not assignable to `JsonString<bigint>`

  JsonStringify and JsonParse
    expected usage
      supports primitive types
        ✔ `boolean`
        ...
        ✔ branded `string`
        ✔ `JsonString<string>`
        ✔ `JsonString<{ arrayOfNumbers: number[] }>`
        ✔ `JsonString<Record<string, number>>`
        ✔ `JsonString<Record<string, number | undefined>>`
        ✔ `JsonString<bigint>`
        ✔ `JsonString<unknown>`
      supports literal types
        ✔ `true`
        ✔ `false`
        ✔ `0`
        ✔ "string"
        ✔ `null`
        ✔ object with literals
        ✔ array of literals
        ✔ tuple of literals
        ✔ specific numeric enum value
        ✔ specific string enum value
        ✔ specific const heterogenous enum value
        ✔ specific computed enum value
      supports array types
        ✔ array of `number`s
        ✔ readonly array of `number`s
        ✔ readonly array of simple objects
      supports object types
        ✔ empty object
        ✔ object with `never`
        ✔ object with `boolean`
        ...
        ✔ `string`|`number` indexed record of `string`s
        ✔ `string`|`number` indexed record of objects
        ✔ templated record of `numbers`
        ✔ `string` indexed record of `number`|`string`s with known properties
        ✔ `string`|`number` indexed record of `strings` with known `number` property (unassignable)
        ✔ branded-`string` indexed of `boolean`s
        ...
        ✔ object with possible type recursion through union
        ✔ object with optional type recursion
        ✔ object with deep type recursion
        ✔ object with alternating type recursion
        ✔ simple non-null object json (NonNullJsonObjectWith<never>)
        ✔ simple read-only non-null object json (ReadonlyNonNullJsonObjectWith<never>)
        ✔ non-const enums
        ✔ object with `readonly`
        ✔ object with getter implemented via value
        ✔ object with setter implemented via value
        ✔ object with matched getter and setter implemented via value
        ✔ object with mismatched getter and setter implemented via value
        class instance
          ✔ with public data (just cares about data)
        object with optional property
          ✔ without property
          ✔ with undefined value
          ✔ with defined value
        opaque Json types
          ✔ opaque serializable object
          ✔ opaque deserialized object
          ✔ opaque serializable and deserialized object
          ✔ opaque serializable unknown
          ✔ opaque deserialized unknown
          ✔ opaque serializable and deserialized unknown
          ✔ object with opaque serializable unknown
          ✔ object with opaque deserialized unknown
          ✔ recursive type with opaque serializable unknown
          ✔ recursive type with opaque deserialized unknown
          ✔ recursive type with opaque serializable and deserialized unknown
          ✔ recursive branded indexed object with OpaqueJsonDeserialized<unknown>
      supports union types
        ✔ simple json (JsonTypeWith<never>)
        ✔ simple read-only json (ReadonlyJsonTypeWith<never>)
      with NOT fully supported object types
        ✔ object with self reference throws on serialization
        known defect expectations
          ✔ sparse array of supported types
          ✔ object with sparse array of supported types
          getters and setters allowed but do not propagate
            ✔ object with `readonly` implemented via getter
            ✔ object with getter
            ✔ object with setter
            ✔ object with matched getter and setter
            ✔ object with mismatched getter and setter
          class instance
            with `ignore-inaccessible-members`
              ✔ with private data ignores private data (that propagates)
    invalid input usage
      assumptions
        ✔ const enums are never readable
      unsupported types cause compiler error
        ✔ `undefined`
        ✔ `unknown`
        ✔ `symbol`
        ✔ `unique symbol`
        ✔ `bigint`
        ✔ function
        ✔ function with supported properties
        ✔ object and function
        ✔ object with function with supported properties
        ✔ object with object and function
        ✔ function with class instance with private data
        ✔ function with class instance with public data
        ✔ class instance with private data and is function
        ✔ class instance with public data and is function
        ✔ `object` (plain object)
        ✔ `void`
        ✔ branded `object`
        ✔ branded object with `string`
        unions with unsupported primitive types
          ✔ `string | symbol`
          ✔ `bigint | string`
          ✔ `bigint | symbol`
          ✔ `number | bigint | symbol`
        array
          ✔ array of `bigint`s
          ✔ array of `symbol`s
          ...
        object
          ✔ object with exactly `bigint`
          ✔ object with optional `bigint`
          ✔ object with exactly `symbol`
          ...
          ✔ `string` indexed record of `unknown`
          ✔ `Partial<>` `string` indexed record of `unknown`
          ✔ `Partial<>` `string` indexed record of `numbers`
          ✔ `Partial<>` templated record of `numbers`
          ✔ object with recursion and `symbol`
          ✔ function object with recursion
          ✔ object and function with recursion
          ✔ nested function object with recursion
          ✔ nested object and function with recursion
          ✔ object with inherited recursion extended with unsupported properties
          object with `undefined`
            ✔ as exact property type
            ✔ in union property
            ...
          object with required `unknown`
            ✔ as exact property type
            ✔ as exact property type adjacent to recursion
            ✔ as exact property type in recursion
          of class instance
            ✔ with private data
            ✔ with private method
            ✔ with private getter
            ✔ with private setter
            ✔ with public method
            ✔ with private data in optional recursion
        opaque Json types requiring extra allowed types
          ✔ opaque serializable object with `bigint`
          ✔ opaque deserialized object with `bigint`
          ✔ opaque serializable and deserialized object with `bigint`
          ✔ opaque serializable object with number array expecting `bigint` support
          ✔ opaque deserialized object with number array expecting `bigint` support
          ✔ opaque serializable and deserialized object with number array expecting `bigint` support
        common class instances
          ✔ Map
          ✔ ReadonlyMap
          ✔ Set
          ✔ ReadonlySet
        Fluid types
          ✔ `IFluidHandle`
          ✔ object with `IFluidHandle`
    special cases
      ✔ explicit `any` generic still limits allowed types
      `number` edge cases
        supported
          ✔ MIN_SAFE_INTEGER
          ✔ MAX_SAFE_INTEGER
          ✔ MIN_VALUE
          ✔ MAX_VALUE
        resulting in `null`
          ✔ NaN
          ✔ +Infinity
          ✔ -Infinity

  JsonDeserialized
    positive compilation tests
      supported primitive types are preserved
        ✔ `JsonString<string>`
        ✔ `JsonString<{ arrayOfNumbers: number[] }>`
        ✔ `JsonString<Record<string, number>>`
        ✔ `JsonString<Record<string, number | undefined>>`
        ✔ JsonString<bigint>
        ✔ JsonString<unknown>

  JsonSerializable
    positive compilation tests
      supported primitive types
        ✔ `JsonString<string>`
        ✔ `JsonString<{ arrayOfNumbers: number[] }>`
        ✔ `JsonString<Record<string, number>>`
        ✔ `JsonString<Record<string, number | undefined>>`
        ✔ JsonString<bigint>
        ✔ JsonString<unknown>
```

jason-ha commented on packages/common/core-interfaces/src/internal.ts:
```diff
@@ -7,9 +7,8 @@
-//  index.js is listed as the runtime file. This is done so that all imports are
-//  using the same outer runtime file. (Could be changed if needed.)
+export type { JsonString, JsonStringifyOptions } from "./jsonString.js";
+export { JsonStringify, JsonParse } from "./jsonString.js";
```
Package cannot be type-only because it has enums.
core-utils does not depend on core-interfaces. If relocated, they could go to client-utils.
Since it can't be type-only and relocation would involve adding test exports to core-interfaces for all of the testing, I am inclined to say this is okay. But I will defer to the prevailing wisdom.

alexvy86 commented on packages/common/core-interfaces/src/jsonString.ts:
```diff
@@ -0,0 +1,75 @@
+ *
+ * - Cast to with `as unknown as JsonString<T>` when value of type `T` has been stringified.
+ *
+ * - Cast from with `as unknown as string` when "instance" will be parsed to `T`.
```
Is this necessary when the type intersection already includes `string`? I'd have expected that to satisfy anything that `as unknown as string` would.

alexvy86 commented on packages/common/core-interfaces/src/jsonString.ts:
```diff
@@ -0,0 +1,75 @@
+	 * When set, inaccessible (protected and private) members throughout type T are
+	 * ignored as if not present.
+	 *
+	 * The default value is not present.
```
```suggestion
	 * The default value is to treat them as if they were not present.
```

jason-ha replied:
I see where there can be confusion here. The default behavior of JsonStringify (when IgnoreInaccessibleMembers is not specified) is to complain if any such properties are present. The default for this setting is that the *setting* is not present. Fixing...

alexvy86 commented on packages/common/core-interfaces/src/test/jsonString.spec.ts:
```diff
@@ -0,0 +1,93 @@
+		);
+	});
+
+	describe("is variant by T", () => {
```
Just reminder about Craig's comment during Hackathon demo meeting, in case the variance descriptions haven't been updated.

alexvy86 commented on packages/common/core-interfaces/src/test/jsonSerializable.spec.ts:
```diff
@@ -1566,8 +1602,10 @@ describe("JsonSerializable", () => {
 				it("`string` indexed record of `unknown`", () => {
-					// @ts-expect-error not assignable to parameter of type '{ [x: string]: JsonTypeWith<never> | OpaqueJsonSerializable<unknown>; }'.
-					const { filteredIn } = passThru(stringRecordOfUnknown);
+					const { filteredIn } = passThru(
+						// @ts-expect-error not assignable to parameter of type '{ [x: string]: JsonTypeWith<never> | OpaqueJsonSerializable<unknown>; }'.
+						stringRecordOfUnknown,
+					);
```
Are we not applying automated formatting to core-interfaces? I don't get what caused the changes from here to the end of the file.

jason-ha replied:
These are manual changes by me to isolate the line that the `ts-expect-error` applies to. Unfortunately, you can't like what error you are expected. When I copied this file and fixed up for JsonStringify and JsonParse testing I noticed these comments were applied too broadly. So not just formatting - opportunistic improvement. And we do apply formatting to all of these.

alexvy86 reviewed (commented):
Haven't looked at `jsonStringifyAndParse.spec.ts‎`. Just minor comments on the rest, and disclaimer that I feel like an ok reviewer but not an expert in this typing black-magic space :).

CraigMacomber commented on packages/common/core-interfaces/src/jsonString.ts:
```diff
@@ -0,0 +1,85 @@
+ *
+ * @sealed
+ */
+declare class JsonStringBrand<T> extends BrandedType<T> {
```
BrandedType lacks documentation for how its intended to be used, so its hard to know it this is using it correctly. Having an example on BrandedType would be good.

jason-ha replied:
To address via #25560

CraigMacomber commented on packages/common/core-interfaces/src/jsonString.ts:
```diff
@@ -0,0 +1,85 @@
+ * @sealed
+ * @internal
+ */
+export type JsonString<T> = string & JsonStringBrand<T>;
```
I'm not convinced this pattern is better than using a different brand for each use case. This gets you structual typing, but nominal might be better.

Suppose for example we have some type like:

`interface Op { metadata: string, inner: unknown }`

And some other interface like:

`interface SummaryNode { metadata: string, inner: SummaryNode, foo: Something }`

I don't think the ability to assign `JsonString<SummaryNode>` to `JsonString<Op>` is a feature. Sure they happen to be structurally compatible, but using less sophisticated branded types for each of them (`type OpString = strng & BrandedType<"OpString">`) would be more semantically robust in my opinion.

If you want both nominal typing and to capture the structure, you could do something like

`declare class JsonStringBrand<T, Name = undefined> extends BrandedType<[T, Name]> {`

or maybe

`type JsonString<T, Name = undefined> = string & BrandedType<[T, "JsonString", Name]> {`

CraigMacomber replied:
Good point: there is no need to couple this stuff with nominal branding: they can be orthogonal.

jason-ha commented on packages/common/core-interfaces/src/jsonString.ts:
```diff
@@ -0,0 +1,85 @@
+ */
+declare class JsonStringBrand<T> extends BrandedType<T> {
+	public toString(): string;
+	private readonly EncodedValue: T;
```
Should be more clear now.

CraigMacomber commented on packages/common/core-interfaces/src/jsonString.ts:
```diff
@@ -0,0 +1,85 @@
+ *
+ * @internal
+ */
+export interface JsonStringifyOptions {
```
Looks like this is duplicated from JsonSerializableOptions?

To avoid duplicating documentation, maybe defined this using `Pick<JsonSerializableOptions, "IgnoreInaccessibleMembers">`, then apply my suggested docs changes to the origional?

CraigMacomber commented on packages/common/core-interfaces/src/test/jsonStringifyAndParse.spec.ts:
```diff
@@ -0,0 +1,2700 @@
+	jsonStringOfStringRecordOfNumberOrUndefined,
+	jsonStringOfBigInt,
+	jsonStringOfUnknown,
+} from "./testValues.js";
```
this is kindof big. Might be a good place to use `import * as testValue ...`

CraigMacomber commented on packages/common/core-interfaces/src/test/jsonStringifyAndParse.spec.ts:
```diff
@@ -0,0 +1,2700 @@
+} from "@fluidframework/core-interfaces/internal/exposedUtilityTypes";
+
+/**
+ * !!! TODO update: Defined using `JsonSerializable` type filter tests `JsonStringify` at call site.
```
I'm not sure what this sentence means, but maybe this todo was supposed to be address in this pr? If not, adjust the phrasing for clarity.

jason-ha committed:
fix(client-core-interfaces): clean-up `@typescript-eslint/ban-types`

jason-ha committed:
docs(client-core-interfaces): complete comment TODO

jason-ha commented on packages/common/core-interfaces/src/jsonString.ts:
```diff
@@ -0,0 +1,85 @@
+ */
+export interface JsonStringifyOptions {
+	/**
+	 * When set, inaccessible (protected and private) members throughout type T are
+	 * ignored as if not present. Otherwise, inaccessible members are considered
+	 * an error (type checking will mention `SerializationErrorPerNonPublicProperties`).
```
This setting shouldn't have anything to do with public properties enumerable or not. But that comment is appropriate for `JsonStringify` assuming that is requirement of `JSON.stringify`.

jason-ha committed:
fix(client-core-interfaces): prevent option bleed through
from JsonStringify to JsonSerializable

jason-ha committed:
docs(client-core-interfaces): update JsonSerializableOptions
and leverage for JsonStringifyOptions while there are no others.

markfields commented on packages/common/core-interfaces/src/test/jsonStringifyAndParse.spec.ts:
```diff
@@ -0,0 +1,2701 @@
+					});
+					assertIdenticalTypes(
+						stringified,
+						createInstanceOf<JsonString<typeof classInstanceWithPublicData>>(),
```
Or this?

```
JsonString<ClassWithPublicData>
```

Also, I'm not sure how to think about `JsonString<SomeClass>` since the conversion is lossy. Meaning, `out: ClassWithPublicData` but... is it really?  What would `out instanceof ClassWithPublicData` say?

markfields replied:
Or just move this case to the "known defects" section

markfields commented on packages/common/core-interfaces/src/test/jsonStringifyAndParse.spec.ts:
```diff
@@ -0,0 +1,2701 @@
+				it("NaN", () => {
+					const { stringified, out } = stringifyThenParse(Number.NaN, null);
+					assertIdenticalTypes(stringified, createInstanceOf<JsonString<number>>());
+					// However, real result is `null`
```
Want to actually assert this?

```suggestion
					assert.equal(out, null, "However, real result is `null`");
```

markfields reviewed (approved):
Thanks for doing such a mind-bogglingly thorough job here!  Looking forward to using this for clarifying op handling code in ContainerRuntime.  Would have helped build confidence in some handle serialization changes we made over a year ago too, if we'd had it.

jason-ha commented on packages/common/core-interfaces/src/test/jsonStringifyAndParse.spec.ts:
```diff
@@ -0,0 +1,2701 @@
+			});
+		});
+
+		describe("supports union types", () => {
```
`JsonTypeWith` and `ReadonlyJsonTypeWith` are a complex union types providing sufficient coverage.

jason-ha commented on packages/common/core-interfaces/src/test/jsonStringifyAndParse.spec.ts:
```diff
@@ -0,0 +1,2701 @@
+					);
+					assertIdenticalTypes(out, classInstanceWithPublicData);
+				});
+				describe("with `ignore-inaccessible-members`", () => {
```
Oversight - they were supposed to be uncommented and fixed up :)

jason-ha committed:
improvement(client-core-interfaces): PR feedback
- Use `as` for `JsonParse` to support `JSON.parse` changing to return `unknown` in future and to be consistent with `JsonStringify` declaration.
- Add test case for "object with optional undefined"
- Cleanup class instance test cases (including those commented out)

jason-ha commented on packages/common/core-interfaces/src/test/jsonStringifyAndParse.spec.ts:
```diff
@@ -0,0 +1,2701 @@
+	ObjectWithOptionalRecursion,
+	ObjectWithSymbolOrRecursion,
+} from "./testValues.js";
+// Note: some values are commented out as not interesting to add coverage for (but acknowledge they exist to test).
```
That would break my use pattern where I diff the list to see what is used or not across some of the tests.
I admit that it feels like there could be something better. It was developed organically with just two files to start and has grown to three files. I don't know what is better but also haven't invested so much time in it.
Per the use pattern it is okay if the list is stale b/c there is a mechanism to check.
See #25596 where I just bring test case sets more in-line.

CraigMacomber replied:
how about you `import * as testValues` then do something like `Object.entries(testValues).filter((t) =>! skipped.has(t[0]))`

markfields replied:
Worth trying IMO, can be in a subsequent PR

jason-ha replied:
I don't quite follow yet but would definitely prefer a follow-up to now.

jason-ha commented on packages/common/core-interfaces/src/test/jsonString.spec.ts:
```diff
@@ -0,0 +1,93 @@
+const jsonStringOfLiteral = JsonStringify("literal");
+
+describe("JsonString", () => {
+	it("`JsonString<string>` is assignable to `string`", () => {
```
Yes. `createInstanceOf` would also show the types more directly. I think the names are decent and IntelliSense is readily available. Sadly, not present in github review view. The other element is that using the values ties things to the runtime tested cases in other related test files.

I do not like the hoops that `requireAssignableTo` requires with allowing unused items and a unique name if you want to have multiple. Or you have to wrap it with something like `function expectTrue<_ extends true>(): void {}`.

So, I'm inclined to keep this as-is but could probably be persuaded.

jason-ha committed:
test(client-core-interface): fix `exactOptionalPropertyTypes:false` test build

jason-ha committed:
improvement(client-core-interfaces): JsonParse parses `JsonString<A> | JsonString<B>` to `JsonDeserialized<A | B>`

jason-ha committed:
improvement(client-core-interfaces): allow `JsonString<A | B>` to be assigned to `JsonString<A> | JsonString<B>`
at the cost of not exactly preserving enum type identities.

jason-ha committed:
fix(client-core-interfaces): restore enum type preservation

jason-ha committed:
test(client-core-interfaces): restore clean test spec

jason-ha committed:
merge: 'main' into core/JsonString

CraigMacomber commented on packages/common/core-interfaces/src/exposedInternalUtilityTypes.ts:
```diff
@@ -903,7 +903,7 @@ export namespace InternalUtilityTypes {
 			? /* test for 'any' */ boolean extends (T extends never ? true : false)
 				? /* 'any' => */ Controls["DegenerateSubstitute"]
-				: Options["IgnoreInaccessibleMembers"] extends "ignore-inaccessible-members"
+				: Options extends { IgnoreInaccessibleMembers: "ignore-inaccessible-members" }
```
Its unclear how somone is supposed to find the docs on JsonSerializableOptions: IgnoreInaccessibleMembers from here.

Up on line 871 I think you should replace

```
{
			IgnoreInaccessibleMembers?: "ignore-inaccessible-members";
		}
```

with `Pick<JsonSerializableOptions, "IgnoreInaccessibleMembers">`

so its clear those are the same field, and that documentation applies.

THen here you could do:

```suggestion
				: Options extends Required<Pick<JsonSerializableOptions, "IgnoreInaccessibleMembers">>
```

jason-ha replied:
That would be a cyclic file import unless that type is relocated, which didn't feel the best for readability for the API file. I am trying to keep the exposed options a bit separated from the internals. See also internal `FilterControls` that is like part of the `JsonSerializableOptions` -- I don't want to have those interdependent.
But, yes, there are no docs now in this place for that internal option. I will add a comment.

markfields commented on packages/common/core-interfaces/src/jsonString.ts:
```diff
@@ -21,6 +22,34 @@ declare class JsonStringBrand<T> extends BrandedType<JsonString<unknown>> {
+ *
+ * The downside is that enums are expanded to union of members and thus cannot be
+ * reconstituted exactly (even if IntelliSense shows the original enum type). This
+ * can be removed if exact enum preservation is found to be more important.
```
I think we like avoiding enums anyway so seems like the right priority

markfields reviewed (approved):
LGTM!

jason-ha committed:
docs(client-core-interfaces): add IgnoreInaccessibleMembers doc reference
